CRM/Core/Form.php

```diff
@@ -2388,66 +2398,115 @@ public static function validateMandatoryFields($fields, $values, &$errors) {
-   *   - contact id validated from a checksum from a checksum
-   *   - cid from the url if the caller has ACL permission to view
-   *   - fallback is logged in user (or ? NULL if no logged in user) (@todo wouldn't 0 be more intuitive?)
+   * Get contact iD for a form object.
+   *
+   * This checks the requestedContactID and returns it if
+   * - it is the number 0 (relevant for online contribution & event forms).
+   * - it is the logged in user
+   * - it is validated by a checksum in the url.
+   * - it is a contact that the logged in user has permission to view
+   *
+   * Failing that it returns the logged in user, if any. This is may be useful
+   * for users taking actions from their contact dashboard (although usually one
+   * of the variants above would be hit).
    *
    * @return NULL|int
+   *
+   * @throws \CRM_Core_Exception
    */
   protected function setContactID() {
-    $tempID = CRM_Utils_Request::retrieve('cid', 'Positive', $this);
-    if (isset($this->_params) && !empty($this->_params['select_contact_id'])) {
-      $tempID = $this->_params['select_contact_id'];
-    }
-    if (isset($this->_params, $this->_params[0]) && !empty($this->_params[0]['select_contact_id'])) {
-      // event form stores as an indexed array, contribution form not so much...
-      $tempID = $this->_params[0]['select_contact_id'];
-    }
+    $requestedContactID = $this->getRequestedContactID();
 
     // force to ignore the authenticated user
-    if ($tempID === '0' || $tempID === 0) {
+    if ($requestedContactID === 0) {
       // we set the cid on the form so that this will be retained for the Confirm page
       // in the multi-page form & prevent us returning the $userID when this is called
       // from that page
       // we don't really need to set it when $tempID is set because the params have that stored
       $this->set('cid', 0);
-      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-      return (int) $tempID;
+      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => $requestedContactID]);
+      return (int) $requestedContactID;
     }
 
-    $userID = CRM_Core_Session::getLoggedInContactID();
-
-    if (!is_null($tempID) && $tempID === $userID) {
-      CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-      return (int) $userID;
-    }
-
-    //check if this is a checksum authentication
-    $userChecksum = CRM_Utils_Request::retrieve('cs', 'String', $this);
-    if ($userChecksum) {
-      //check for anonymous user.
-      $validUser = CRM_Contact_BAO_Contact_Utils::validChecksum($tempID, $userChecksum);
-      if ($validUser) {
-        CRM_Core_Resources::singleton()->addVars('coreForm', ['contact_id' => (int) $tempID]);
-        CRM_Core_Resources::singleton()->addVars('coreForm', ['checksum' => $userChecksum]);
-        return $tempID;
+    if ($requestedContactID === $this->getAuthenticatedContactID()) {
```

Comment:
Note that `$this->getAuthenticatedContactID()` is always an int so will not match NULL here